### ISSUE TITLE
Authenticated method does not return the right status code

### ISSUE THREAD

**Issue by hshahdoost** [Folkloreatelier/laravel-graphql#358]
Please correct me if I'm wrong but I believe the `authentcated` method in `Field` class, (which we can override in every query or mutation to ensure authentication) is not returning the right status code. Since `authorized` method returns stats code of 403 (forbidden) in case of failure. `authentcated` method shall return 401 (unauthenticated). 
Keep in mind that this status code is very important for the client authentication process, since libraries such as `OkHttpClient` consider the 401 authentication failure and would automatically call the authenticator callback when it's being returned.

**Comment by mfn** [comment]
Btw, small technicality: 401 is "unauthorized", not unauthenticated.

**Comment by mfn** [comment]
Take a look at https://github.com/Folkloreatelier/laravel-graphql/blob/171631a17e4bf33780513f15221ea6052c67d498/src/Folklore/GraphQL/GraphQLController.php#L72-L77 

If any error message matches `Unauthorized`, it's automatically turned into 403.

You can extend the existing controller and use your own, see https://github.com/Folkloreatelier/laravel-graphql/blob/171631a17e4bf33780513f15221ea6052c67d498/src/config/config.php#L49

This way you can implement your own logic to return 401 for now.

**Comment by hshahdoost** [comment]
@mfn Yes, I'm sorry, I actually wrote the word incorrectly, it was **Unauthorized**.
but, as you better know the difference between forbidden (403) and unauthorized (401) is that when 401 is returned it is expected to be solved by authentication (at the first level) and then authorization. But receiving 403 means that there is no way we can access to this data, so the request should not be repeated. (as being explained [here](https://www.w3.org/Protocols/rfc2616/rfc2616-sec10.html#sec10.4.2))
Now all I'm saying is that It would be more appropriate to receive 401 since the client side considers it as a valid status code for authentication.
In the meantime, thanks a lot for such a quick response.

**PS**: I also saw that part of the code, and wondered wouldn't it be better if the status code of the page would be set in the `Error` class? I know that graphql is on top of http and should not involve with it's technicality, but I think somehow it would be nicer to be able to extend the `Error` class and implement different types of errors with different status codes.

**Comment by mfn** [comment]
> but I think somehow it would be nicer to be able to extend the Error class and implement different types of errors with different status codes

I also dislike the pattern matching in the errors and changing the error code based on this.

I would find it more natural to react depending on the type of Exception/Error thrown, alas this is not how the library currently works 🤷‍♀️

What I did:
- I use a custom error formatter
- I inspect the error/exception and decide whether to _rethrow_ it
- in which case it's handled by your apps `Exceptions\Handler`
- where you can now return what you want

Or maybe it would be easier to just use a custom GraphQL controller etc.

One reason for going down the rethrow route: by default this library swallows _all exceptions_ which is a problem if you use 3rd party error handlers like Sentry. If e.g. an database error occurs, you want to receive that in your central error facility.

**Comment by hshahdoost** [comment]
@mfn Thanks a lot for your response. Well I actually wrote this because I thought it could be useful. Any of the solutions you gave are fine, and would solve the problem. I personally solved it by defining two schema, one for guest user and the other for authenticated users so that way everything was handled by laravelpassport.
